internal/kafka/internal/services/cluster_placement_strategy.go

```diff
@@ -129,3 +131,52 @@ func (f *FirstSchedulableWithinLimit) findClusterKafkaInstanceCount(clusterIDs [
 		return clusterWithinLimitMap, nil
 	}
 }
+
+// FirstReadyWithCapacity finds and returns the first cluster in a Ready status with remaining capacity
+type FirstReadyWithCapacity struct {
+	ClusterService ClusterService
+	KafkaConfig    *config.KafkaConfig
+}
+
+func (f *FirstReadyWithCapacity) FindCluster(kafka *dbapi.KafkaRequest) (*api.Cluster, *errors.ServiceError) {
```

Comment:
we should return `error` instead of `*errors.ServiceError`

Reply:
when this is changed to `error`, should the underlying error types returned from the functions these strategies call also be changed to type `error`?

Reply:
All functions should return `error` unless there is a real strong reason to do otherwise IMO.

It is caller duty to cast the `error` interface to the concrete type if needed.

Reply:
I think this may be a substantial change. I'm happy to follow up on this in a separate PR, to make it easier for people to review this one. wdyt?

Reply:
I'd expect that most of the places where this function is called won't do much more than

```go
if err != nil {
   return nil
}
```

Unless there are many places that are directly referring to the `ServiceError` type when invoking this methods, the change shouldn't be too far from just changing the return to `error`.

That being said, I'm ok with doing this in a follow up PR: the reason I pointed it out here was that I think we should start following the _return error rule_ in the newly added functions.

Reply:
May not too bad in the service itself but I think most of the changes will be on the tests.